diff --git a/dlls/ncrypt/main.c b/dlls/ncrypt/main.c
index 03ff94e1d14..3d573de2d33 100644
--- a/dlls/ncrypt/main.c
+++ b/dlls/ncrypt/main.c
@@ -160,6 +160,14 @@ static struct object *create_key_object(enum algid algid, NCRYPT_PROV_HANDLE pro
         if (!(object = allocate_object(KEY))) return NULL;
 
         object->key.algid = ECDSA_P256;
+        set_object_property(object, NCRYPT_ALGORITHM_PROPERTY, (BYTE *)BCRYPT_ECDSA_P256_ALGORITHM,
+                            sizeof(BCRYPT_ECDSA_P256_ALGORITHM));
+        set_object_property(object, NCRYPT_ALGORITHM_GROUP_PROPERTY, (BYTE *)BCRYPT_ECDSA_ALGORITHM,
+                            sizeof(BCRYPT_ECDSA_ALGORITHM));
+        dw_value = 256;
+        set_object_property(object, NCRYPT_LENGTH_PROPERTY, (BYTE *)&dw_value, sizeof(dw_value));
+        dw_value = 64;
+        set_object_property(object, BCRYPT_SIGNATURE_LENGTH, (BYTE *)&dw_value, sizeof(dw_value));
         break;
 
     default:
@@ -460,6 +468,33 @@ SECURITY_STATUS WINAPI NCryptImportKey(NCRYPT_PROV_HANDLE provider, NCRYPT_KEY_H
         set_object_property(object, BCRYPT_PUBLIC_KEY_LENGTH, (BYTE *)&rsablob->BitLength, sizeof(rsablob->BitLength));
         break;
     }
+    case BCRYPT_ECDSA_PRIVATE_P256_MAGIC:
+    case BCRYPT_ECDSA_PUBLIC_P256_MAGIC:
+    {
+        NTSTATUS status;
+        BCRYPT_ECCKEY_BLOB *eccblob = (BCRYPT_ECCKEY_BLOB *)data;
+        DWORD length;
+
+        if (!(object = create_key_object(ECDSA_P256, provider)))
+        {
+            ERR("Error allocating memory\n");
+            return NTE_NO_MEMORY;
+        }
+
+        status = BCryptImportKeyPair(BCRYPT_ECDSA_P256_ALG_HANDLE, NULL, type, &object->key.bcrypt_key, data,
+                                     datasize, 0);
+        if (status != STATUS_SUCCESS)
+        {
+            WARN("Error importing key pair %#lx\n", status);
+            free(object);
+            return map_ntstatus(status);
+        }
+
+        length = eccblob->cbKey * 8;
+        set_object_property(object, NCRYPT_LENGTH_PROPERTY, (BYTE *)&length, sizeof(length));
+        set_object_property(object, BCRYPT_PUBLIC_KEY_LENGTH, (BYTE *)&length, sizeof(length));
+        break;
+    }
     default:
         FIXME("Unhandled key magic %#lx\n", header->Magic);
         return NTE_INVALID_PARAMETER;
